File: litedb-java/src/main/java/com/litedb/cluster/RejoinSmoke.java
package com.litedb.cluster;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * RejoinSmoke — proves a reaped node that REJOINS doesn't disrupt the cluster (the Pre-Vote fix;
 * mirror of rejoin_smoke.py). Kill the PD leader (a node), let auto-heal reap it, then RESTART and
 * re-add it. The returning node still has stale shard replicas on disk; WITHOUT pre-vote they keep
 * forcing higher-term elections and leave shards leaderless forever. WITH pre-vote, peers refuse the
 * stale replica's pre-vote while hearing from their leader, so the cluster converges back to fully
 * HEALTHY and stays there.
 *
 * Run with 4 nodes:  LITEDB_CLUSTER_NODES=4 java com.litedb.cluster.RejoinSmoke
 */
public class RejoinSmoke {

    static Map<String, Process> procs = new LinkedHashMap<>();

    static Process spawn(String nid) throws Exception {
        ProcessBuilder pb = new ProcessBuilder(System.getProperty("java.home") + "/bin/java",
                "-cp", System.getProperty("java.class.path"), "com.litedb.cluster.NodeServer", nid);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);
        Process p = pb.start();
        procs.put(nid, p);
        return p;
    }

    @SuppressWarnings("unchecked")
    static Set<String> readyLeaders(ClusterClient client) {
        Set<String> out = new java.util.HashSet<>();
        for (Map<String, Object> st : client.status()) {
            if (st.get("shards") == null) continue;
            for (Object so : (List<Object>) st.get("shards")) {
                Map<String, Object> sh = (Map<String, Object>) so;
                if ("leader".equals(sh.get("role")) && Boolean.TRUE.equals(sh.get("ready"))) {
                    out.add((String) sh.get("group"));
                }
            }
        }
        return out;
    }

    @SuppressWarnings("unchecked")
    static String pdLeader(Rpc.Client rpc) {
        for (String n : ClusterConfig.PD_NODES) {
            Map<String, Object> r = rpc.call(ClusterConfig.HOST, ClusterConfig.port(n), "pd_status", NodeServer.resp());
            if (Boolean.TRUE.equals(r.get("ok")) && r.get("result") instanceof Map) {
                Map<String, Object> res = (Map<String, Object>) r.get("result");
                if (res.get("leader") != null) return (String) res.get("leader");
            }
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    static Set<String> hostedBy(ClusterClient client, String node) {
        Set<String> out = new java.util.HashSet<>();
        for (Map<String, Object> st : client.status()) {
            if (!node.equals(st.get("node")) || st.get("shards") == null) continue;
            for (Object so : (List<Object>) st.get("shards")) out.add((String) ((Map<String, Object>) so).get("group"));
        }
        return out;
    }

    static void waitUntil(java.util.function.BooleanSupplier c, long ms, String what) throws Exception {
        long deadline = System.currentTimeMillis() + ms;
        while (System.currentTimeMillis() < deadline) { if (c.getAsBoolean()) return; Thread.sleep(250); }
        throw new AssertionError("timeout: " + what);
    }

    public static void main(String[] args) throws Exception {
        int nShards = ClusterConfig.SHARDS.size();
        Path root = Paths.get(ClusterConfig.dataRoot());
        if (Files.exists(root)) Files.walk(root).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
        Controller ctrl = new Controller(ClusterConfig.INITIAL_NODES, System.out::println);
        Rpc.Client rpc = new Rpc.Client(1500);
        try {
            for (String n : ClusterConfig.INITIAL_NODES) spawn(n);
            ClusterClient client = new ClusterClient();
            waitUntil(() -> readyLeaders(client).size() == nShards, 30000, "all shards have a leader at start");
            for (int i = 0; i < 12; i++) if (!Boolean.TRUE.equals(client.put("key" + i, "val" + i).get("ok"))) throw new AssertionError("put key" + i);
            String victim = pdLeader(rpc);
            if (victim == null) victim = "node-1";
            System.out.println("cluster healthy; PD leader = " + victim + ". Killing it (a node + the PD leader)...");

            final String dead = victim;
            procs.get(dead).destroyForcibly();
            procs.get(dead).waitFor();
            waitUntil(() -> !ctrl.active().contains(dead) && readyLeaders(client).size() == nShards,
                    70000, "auto-heal after the PD leader died (reaped + RF restored)");
            System.out.println("  auto-healed: " + dead + " reaped, all " + nShards + " shards led on " + ctrl.active());

            System.out.println("restarting " + dead + " and re-adding it (it still has stale shard replicas on disk)...");
            spawn(dead);
            Thread.sleep(2000);
            ctrl.addNode(dead);

            // THE FIX: the rejoining node must CONVERGE to a stable, fully-led cluster. Reconfiguration
            // churn can briefly dip a shard's leader, so we require SUSTAINED health (5 consecutive
            // seconds). Without pre-vote the stale replicas disrupt elections and this never converges.
            int streak = 0;
            long deadline = System.currentTimeMillis() + 90000;
            while (System.currentTimeMillis() < deadline) {
                streak = readyLeaders(client).size() == nShards ? streak + 1 : 0;
                if (streak >= 5) break;
                Thread.sleep(1000);
            }
            if (streak < 5) throw new AssertionError("cluster never stabilized after rejoin — a returning node is disrupting elections");
            for (int i = 0; i < 12; i++) if (!("val" + i).equals(client.get("key" + i))) throw new AssertionError("data lost after rejoin");

            // FENCING: the rejoined node must host only shards it's a voter of — its orphaned stale
            // replicas (shards it was reaped from) must have been dropped + wiped, not left lingering.
            Thread.sleep(6000);
            Map<String, List<String>> placement = ctrl.placement();
            Set<String> hosted = hostedBy(client, dead);
            for (String s : hosted) {
                if (!placement.getOrDefault(s, List.of()).contains(dead)) {
                    throw new AssertionError(dead + " still hosts orphan replica of " + s + " it isn't a voter of");
                }
            }
            System.out.println("  " + dead + " hosts only shards it's a voter of " + hosted + " — orphans fenced + wiped");

            System.out.println("\nREJOIN OK (Java): " + dead + " restarted + re-added with stale replicas did NOT "
                    + "disrupt the cluster (pre-vote refused its elections) — all " + nShards + " shards keep a "
                    + "stable leader, data intact.");
        } finally {
            ctrl.stop();
            for (Process p : procs.values()) p.destroyForcibly();
            for (Process p : procs.values()) p.waitFor();
        }
        System.exit(0);
    }
}

File: litedb-python/rejoin_smoke.py
"""
rejoin_smoke.py — proves a reaped node that REJOINS doesn't disrupt the cluster (the Pre-Vote fix).

Reproduces the failure we hit by hand: kill the PD leader (a node), let auto-heal reap it, then
RESTART and re-add it. The returning node still has stale shard replicas on disk (its drop_shard was
skipped while it was dead), so it tries to resurrect them. WITHOUT pre-vote, those stale replicas keep
starting higher-term elections and disrupt the healthy groups → some shards lose their leader forever
(DEGRADED). WITH pre-vote, peers refuse the stale replica's pre-vote while they're hearing from their
leader, so it can't bump the term — the cluster returns to fully HEALTHY and stays there.

Run:  python rejoin_smoke.py
"""

import os

os.environ.setdefault("LITEDB_CLUSTER_NODES", "4")

import shutil  # noqa: E402
import subprocess  # noqa: E402
import sys  # noqa: E402
import time  # noqa: E402

import _loader  # noqa: F401,E402
from cluster_client import ClusterClient  # noqa: E402
from cluster_config import DATA_ROOT, INITIAL_NODES, NODES, PD_NODES, SHARDS  # noqa: E402
from controller import Controller  # noqa: E402
from rpc import RPCClient  # noqa: E402


def spawn(nid: str):
    return subprocess.Popen([sys.executable, "node.py", nid],
                            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def wait_until(fn, timeout=70.0, what=""):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if fn():
            return True
        time.sleep(0.25)
    raise AssertionError("timeout: " + what)


def ready_leaders(client) -> set:
    out = set()
    for st in client.status():
        for sh in st.get("shards", []):
            if sh["role"] == "leader" and sh.get("ready"):
                out.add(sh["group"])
    return out


def main():
    n_shards = len(SHARDS)
    shutil.rmtree(DATA_ROOT, ignore_errors=True)
    procs = {n: spawn(n) for n in INITIAL_NODES}
    ctrl = Controller()
    rpc = RPCClient(timeout=1.5)

    def pd_leader():
        for n in PD_NODES:
            host, port = NODES[n]
            r = rpc.call(host, port, "pd_status", {})
            res = r.get("result") if r.get("ok") else None
            if isinstance(res, dict) and res.get("leader"):
                return res["leader"]
        return None

    try:
        client = ClusterClient()
        wait_until(lambda: len(ready_leaders(client)) == n_shards, what="all shards have a leader at start")
        for i in range(12):
            assert client.put(f"key{i}", f"val{i}").get("ok"), f"put key{i}"
        victim = pd_leader() or "node-1"
        print(f"cluster healthy; PD leader = {victim}. Killing it (a node + the PD leader)...")

        procs[victim].kill()
        procs[victim].wait()
        wait_until(lambda: victim not in ctrl.active and len(ready_leaders(client)) == n_shards,
                   what="auto-heal after the PD leader died (reaped + RF restored)")
        print(f"  auto-healed: {victim} reaped, all {n_shards} shards led on survivors {sorted(ctrl.active)}")

        print(f"restarting {victim} and re-adding it (it still has stale shard replicas on disk)...")
        procs[victim] = spawn(victim)
        time.sleep(2)
        ctrl.add_node(victim)

        # THE FIX: the rejoining node must CONVERGE to a stable, fully-led cluster. Reconfiguration
        # churn (node-1 is re-added to several shards) can briefly dip a shard's leader, so we wait for
        # SUSTAINED health — 5 consecutive seconds of all shards led. Without pre-vote the stale
        # replicas disrupt elections permanently and this never converges (it stays stuck, as we saw).
        def stably_healthy(streak_needed=5, timeout=80.0) -> bool:
            streak = 0
            deadline = time.monotonic() + timeout
            while time.monotonic() < deadline:
                if len(ready_leaders(client)) == n_shards:
                    streak += 1
                    if streak >= streak_needed:
                        return True
                else:
                    streak = 0
                time.sleep(1.0)
            return False

        assert stably_healthy(), \
            "cluster never stabilized after rejoin — a returning node is disrupting elections"
        assert all(client.get(f"key{i}") == f"val{i}" for i in range(12)), "data intact after rejoin"

        # FENCING: the rejoined node must host only shards it's actually a voter of — its orphaned
        # stale replicas (shards it was reaped from) must have been dropped + wiped, not left lingering.
        time.sleep(6)  # let the fence loop run
        placement = ctrl.placement
        hosted = {sh["group"] for st in client.status() if st.get("node") == victim
                  for sh in st.get("shards", [])}
        orphans = {s for s in hosted if victim not in placement.get(s, [])}
        assert not orphans, f"{victim} still hosts orphan replicas it isn't a voter of: {orphans}"
        print(f"  {victim} hosts only shards it's a voter of {sorted(hosted)} — orphans fenced + wiped")

        print(f"\nREJOIN OK: {victim} restarted + re-added with stale replicas did NOT disrupt the "
              f"cluster (pre-vote refused its elections) — all {n_shards} shards keep a stable leader, "
              f"data intact.")
    finally:
        ctrl.stop()
        for p in procs.values():
            if p.poll() is None:
                p.kill()
        for p in procs.values():
            p.wait()


if __name__ == "__main__":
    main()
